server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -537,6 +539,10 @@ private boolean shouldRefresh() {
 
     @Override
     public ClusterInfo getClusterInfo() {
+        return currentClusterInfo;
+    }
+
+    private ClusterInfo updateAndGetCurrentClusterInfo() {
```

Comment:
As discussed in chat, the point of separating get and update is to remove the risk of someone calling get when a refresh is half-finished and receiving a `ClusterInfo` that was a mix of data from the current and previous refresh.

I wonder in light of this change whether we should make

`leastAvailableSpaceUsages` / `mostAvailableSpaceUsages` / `org.elasticsearch.cluster.InternalClusterInfoService#nodeThreadPoolUsageStatsPerNode` etc. all fields on the `AsyncRefresh`, so that state is kept private until the refresh was completed? Feels a bit safer than leaving it as fields on the enclosing class where it can be accessed whenever.

If not that we should at least put some doc here to indicate that using any of those fields directly isn't safe and the reason for the update/get split.

Reply:
It would mean rather than `updateAndGet...` living at the top level and accessing fields, it might need to change to being `updateClusterInfo` and accepting one that the `AsyncRefresh` had created from its private state.

Reply:
I'd have a separate PR for the refactoring. For now I added comments in 95ba02db38415d6fb4e58471393b31d876b4bb12